### DISCUSSION TITLE
Next, SSR, authExchange and cookie auth

### DISCUSSION THREAD

**Discussion by saevarb** [urql-graphql/urql#1973] [Q&A] [answered]
Hello.

I have built a nextjs application, and that same server hosts a GQL API which I am querying with urql. Authentication is done via an HTTP-only cookie, and I'm using next-urql.

I also want to use `getServerSideProps` to authenticate my pages, in case that is relevant.

My question is: How do I properly configure `authExchange` (and I guess any other relevant part of urql) so that urql can run during SSR when using cookie auth? It seems to me that on the server, I need to be able to access the next context's request and grab the cookie, so we can use it on the user's behalf. When on the client, nothing special needs to be done because the cookie is send automatically.

I thought I might be able to do this using the `context` parameter given by `withUrqlClient`, but it seems that this is always undefined on the server.

**Comment by JoviDeCroock** [comment, marked as answer]
`context` is injected when using the `ssr: true` option which utilizes `react-ssr-prepass` and `getInitialProps` to aggregate data upfront, if you need the context in `getServerSideProps` you'll have to incorporate the passing of said context in the server-side function like this:

```js
export async function getStaticProps(ctx) {
  const ssrCache = ssrExchange({ isClient: false });
  const client = initUrqlClient({
    url: 'your-url',
    // pass in the ctx from the upper function in here
    exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
  });

  await client.query(TODOS_QUERY).toPromise();

  return {
    props: {
      urqlState: ssrCache.extractData(),
    },
    revalidate: 600,
  };
}
```

**Reply by edw19** [reply]
as context is passed, don't see that property in initUrqlCLient?

![image](https://user-images.githubusercontent.com/47750859/136482888-7e4251d5-08be-48d1-a499-ba217ecfa161.png)

**Reply by Axedyson** [reply]
Would this be a proper way to execute multiple queries during SSR?:
```TypeScript
export async function getStaticProps(ctx) {
  const ssrCache = ssrExchange({ isClient: false });
  const client = initUrqlClient({
    url: 'your-url',
    exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
  });

  // Is this fine? Or is there a better way? maybe some function to run multiple queries in parallel?
  for (const query of queries) await client!.query(query).toPromise()

  return {
    props: {
      // urqlState is a keyword here so withUrqlClient can pick it up.
      urqlState: ssrCache.extractData(),
    },
    revalidate: 600,
  };
}
```
